meteor/client/ui/Settings/ShowStyle/VariantSettings.tsx

```diff
@@ -30,22 +44,66 @@ interface IShowStyleVariantsProps {
 interface IShowStyleVariantsSettingsState {
 	editedMappings: ProtectedString<any>[]
 	timestampedFileKey: number
+	dndVariants: ShowStyleVariant[]
 }
 
+interface DragVariant {
+	index: number
+	type: ShowStyleDragDropTypes
+}
+
+const SET_STATE_DELAY = 50
+
 export const ShowStyleVariantsSettings = withTranslation()(
 	class ShowStyleVariantsSettings extends React.Component<
 		Translated<IShowStyleVariantsProps>,
 		IShowStyleVariantsSettingsState
 	> {
+		private timer?: number
+
 		constructor(props: Translated<IShowStyleVariantsProps>) {
 			super(props)
 
 			this.state = {
 				editedMappings: [],
 				timestampedFileKey: Date.now(),
+				dndVariants: this.props.showStyleVariants,
+			}
+		}
+		componentDidUpdate(prevProps: Readonly<Translated<IShowStyleVariantsProps>>) {
+			if (this.showStyleVariantsChanged(prevProps) || this.noShowStyleVariantsPresentInState()) {
+				this.timer = Meteor.setTimeout(() => {
+					this.setState({
+						dndVariants: this.props.showStyleVariants,
+					})
+				}, SET_STATE_DELAY)
+			} else if (this.timer) {
+				Meteor.clearTimeout(this.timer)
 			}
```

Comment:
I don't entirely understand how this is supposed to work, so maybe I misunderstand it, but what if someone reorders the variants (the props change; `this.showStyleVariantsChanged(prevProps)` returns `true` and the timeout is set), but within 50ms some other prop change arrives, not necessarily affecting the variants, leading to `this.showStyleVariantsChanged(prevProps) || this.noShowStyleVariantsPresentInState()` returning `false` and causing the timeout to be cleared, preventing the new order from being shown?

Reply:
The way i understand it, the specified methods only await execution, the line of methods to be executed is not halted in terms of accepting new methods to run until it is finished with the ongoing timeout. I might not fully understand it myself, but in any case i've tested with a very high timeout and made many different request to check if anything gets lost, and nothing has happened so far... If you come up with a different answer please let me know.

Reply:
So here's how this looks after setting the delay to a _very high_ 1000ms. I added log lines (show below) to show you what's happening. Any prop updates that don't fall into the first condition in `componentDidUpdate`, will clear the existing timer and block updates. It is not safe to assume that within 50ms, or frankly any amount of time `componentDidUpdate` will not be called again for _some reason_.


https://user-images.githubusercontent.com/9103996/203258739-895acfa0-18aa-4f3f-a3f3-98daeb15f475.mp4
``` typescript
		componentDidUpdate(prevProps: Readonly<Translated<IShowStyleVariantsProps>>) {
			if (this.showStyleVariantsChanged(prevProps) || this.noShowStyleVariantsPresentInState()) {
				console.log('set timeout :)')
				this.timer = Meteor.setTimeout(() => {
					console.log('set state :D')
					this.setState({
						dndVariants: this.props.showStyleVariants,
					})
				}, SET_STATE_DELAY)
			} else if (this.timer) {
				console.log('cleared timeout :(')
				Meteor.clearTimeout(this.timer)
			}
		}
```

Reply:
With the newest commit the timeout no longer clears in the update method to avoid this scenario :-)